src/lang/base/Syntax.ml

```diff
@@ -747,39 +744,53 @@ module ScillaSyntax (SR : Rep) (ER : Rep) = struct
           List.fold plist ~init:acc ~f:(fun acc p' -> accfunc p' acc)
       in accfunc p []
     in
-
+  
+    (* The main function that does the job. *)
     let rec recurser erep bound_vars =
       let (e, _) = erep in
       match e with
-      | Literal _ -> false
-      | Var v -> (not @@ is_mem v bound_vars) && is_mem v blist
-      | Fun (f, _, body) -> recurser body (f :: bound_vars)
+      | Literal _ -> []
+      | Var v | TApp (v, _) -> if is_mem v bound_vars then [] else [v]
+      | Fun (f, _, body) | Fixpoint (f, _, body) -> recurser body (f :: bound_vars)
       | TFun (_, body) -> recurser body bound_vars
-      | Constr (_, _, es) -> any_is_mem (get_free es bound_vars) blist
-      | App (f, args) ->
-        let args' = f :: args in
-        any_is_mem (get_free args' bound_vars) blist
-      | Builtin (_f, args) ->
-        any_is_mem (get_free args bound_vars) blist
+      | Constr (_, _, es) -> get_free es bound_vars
+      | App (f, args) -> get_free (f :: args) bound_vars
+      | Builtin (_f, args) -> get_free args bound_vars
       | Let (i, _, lhs, rhs) ->
-        (recurser lhs bound_vars) || (recurser rhs (i::bound_vars))
+        let fv_lhs = recurser lhs bound_vars in
+        let fv_rhs = recurser rhs (i::bound_vars) in
+       (fv_lhs @ fv_rhs)
```

Comment:
This bit causes the function to be quadratic.

If you add an accumulator to the recurser, you can get it to be linear:

```
let rec recurser erep bound_vars acc =
  ...
  match e with
  | Literal _ -> acc
  | Var v -> if is_mem v bound_vars then acc else v :: acc
  ...
  | Let (i, _, lhs, rhs) ->
    let acc_lhs = recurser lhs bound_vars acc in
    recurser rhs (i::bound_vars) acc_lhs
  ...
```

and so on.

Reply:
I'll do this change and update.

Reply:
@jjcnn I've made the change, can you have a quick look? Thank you.